frotz/src/interface/frotz_interface.c

```diff
@@ -1699,3 +1700,127 @@ void test() {
     printf("Attr Clr %d: %d --> %d\n", i, attr_clr_objs[i], attr_clr_nb[i]);
   }
 }
+
+// Given a list of action candidates, find the ones that lead to valid world changes.
+// candidate_actions contains a string with all the candidate actions, seperated by ';'
+// valid_actions will be written with each of the identified valid actions seperated by ';'
+// equiv_classes will be written with the world_diff index for each valid_action indicating
+// which of the valid actions are equivalent to each other in terms of their world diffs.
+// Returns the number of valid actions found.
+int filter_candidate_actions(char *candidate_actions, char *valid_actions, unsigned char *equiv_classes) {
+  char *act = NULL;
+  char *act_newline = NULL;
+  char *text;
+  short orig_score;
+  int i;
+  int valid_cnt = 0;
+  int v_idx = 0;
+  
+  // Variables used to store & reset game state
+  unsigned char ram_cpy[h_dynamic_size];
+  unsigned char stack_cpy[STACK_SIZE*sizeof(zword)];
+  int pc_cpy;
+  int sp_cpy;
+  int fp_cpy;
+  int frame_count_cpy;
+  long rngA_cpy;
+  int rngInterval_cpy;
+  int rngCounter_cpy;
+
+  int diff_arr_size = 100;
+  zword diff[128];
+  zword diff_array[128*diff_arr_size];
+  int diff_array_idx = 0;
+  int match_idx;
+
+  // Save the game state
+  getRAM(ram_cpy);
+  getStack(stack_cpy);
+  pc_cpy = getPC();
+  sp_cpy = getSP();
+  fp_cpy = getFP();
+  frame_count_cpy = getFrameCount();
+  rngA_cpy = getRngA();
+  rngInterval_cpy = getRngInterval();
+  rngCounter_cpy = getRngCounter();
+
+  orig_score = get_score();
+  // zero the diff array
+  memset(diff_array, 0, 128*diff_arr_size*sizeof(zword));
+
+  act = strtok(candidate_actions, ";");
+  while (act != NULL)
+  {
+    // Add a newline and termination to act
+    act_newline = malloc(strlen(act) + 2);
+    strcpy(act_newline, act);
+    strcat(act_newline, "\n");
+
+    // Step: Code is copied due to inexplicable segfault when calling step() directly; Ugh!
```

Comment:
Do you think it was the same issue as with the `opcode` (not sure why it would be the case, though)?

Reply:
I don't think so - it was a segfault and happens on games like Zork which never had the illegal opcode issue.

Reply:
ok